metadata/src/test/java/org/apache/kafka/controller/LocalLogManagerTest.java

```diff
@@ -0,0 +1,258 @@
+                LeaderInfo next = env.waitForLeader();
+                while (next.epoch() == cur.epoch()) {
+                    Thread.sleep(1);
+                    next = env.waitForLeader();
+                }
+                long expectedNextEpoch = cur.epoch() + 1;
+                assertEquals("Expected next epoch to be " + expectedNextEpoch +
+                    ", but found  " + next, expectedNextEpoch, next.epoch());
+                cur = next;
+            } while (cur.nodeId() == first.nodeId());
+            env.close();
+            assertEquals(null, env.firstError.get());
+        }
+    }
+
+    static class TestCommit {
+        final long epoch;
+        final long index;
+        final ApiMessage message;
+
+        public TestCommit(long epoch, long index, ApiMessage message) {
+            this.epoch = epoch;
+            this.index = index;
+            this.message = message;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(epoch, index, message);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof TestCommit)) {
+                return false;
+            }
+            TestCommit other = (TestCommit) o;
+            return epoch == other.epoch &&
+                index == other.index &&
+                message.equals(other.message);
+        }
+
+        @Override
+        public String toString() {
+            return "TestCommit(epoch=" + epoch + ", index=" + index +
+                ", message=" + message + ")";
+        }
+    }
+
+    static class CommitChecker implements CommitHandler {
+        private final List<TestCommit> commits;
+        private final Map<Integer, Integer> cursors = new HashMap<>();
+
+        CommitChecker(List<TestCommit> commits) {
+            this.commits = commits;
+        }
+
+        @Override
+        public synchronized void handle(int nodeId, long epoch, long index,
+                                        ApiMessage message) throws Exception {
+            int commitNumber = cursors.getOrDefault(nodeId, 0);
+            TestCommit newCommit = new TestCommit(epoch, index, message);
+            TestCommit expectedCommit = commits.get(commitNumber);
+            assertEquals("Unexpected commit on node " + nodeId, expectedCommit, newCommit);
+            cursors.put(nodeId, commitNumber + 1);
+        }
+
+        synchronized void verifyCursorsHaveReachedCommitNumber(int expectedNumCursors,
+                                                               int expectedCommitNumber) {
+            if (cursors.size() < expectedNumCursors) {
+                throw new RuntimeException("Expected " + expectedNumCursors +
+                    " cursors, but there were only " + cursors.size());
+            }
+            for (Map.Entry<Integer, Integer> entry : cursors.entrySet()) {
+                if (entry.getValue() < expectedCommitNumber) {
```

Comment:
!=

Reply:
This one is OK.  It's all right to have committed past the designated index.